library/src/cache/patch_manager.rs

```diff
@@ -538,13 +580,50 @@ mod debug_tests {
         let temp_dir = TempDir::new("patch_manager").unwrap();
         let patch_manager = PatchManager::new(temp_dir.path().to_owned(), Some("public_key"));
         let expected_str = format!(
-            "PatchManager {{ root_dir: \"{}\", patches_state: PatchesState {{ last_booted_patch: None, last_attempted_patch: None, next_boot_patch: None, highest_seen_patch_number: None }}, patch_public_key: Some(\"public_key\") }}",
+            "PatchManager {{ root_dir: \"{}\", patches_state: PatchesState {{ last_booted_patch: None, last_attempted_patch: None, next_boot_patch: None, currently_booting_patch: None, highest_seen_patch_number: None }}, patch_public_key: Some(\"public_key\") }}",
             temp_dir.path().display()
         );
         assert_eq!(format!("{:?}", patch_manager), expected_str);
     }
 }
 
+#[cfg(test)]
+mod on_init_tests {
+    use super::*;
+
+    #[test]
+    fn clears_currently_booting_patch() -> Result<()> {
+        let temp_dir = TempDir::new("patch_manager").unwrap();
+        let mut manager = PatchManager::manager_for_test(&temp_dir);
+
+        // Add a patch and start to boot from it.
+        manager.add_patch_for_test(&temp_dir, 1)?;
+        manager.record_boot_start_for_patch(1)?;
+
+        assert_eq!(
+            manager
+                .patches_state
+                .currently_booting_patch
+                .as_ref()
+                .map(|p| p.number),
+            Some(1)
+        );
+        assert_eq!(
+            manager.next_boot_patch().as_ref().map(|p| p.number),
+            Some(1)
+        );
+
+        // Simulate that the app is being started fresh (e.g. from a crash)
+        manager = PatchManager::manager_for_test(&temp_dir);
+        manager.on_init()?;
+
+        // Verify that we are no longer booting from patch 1.
+        assert!(manager.next_boot_patch().is_none());
```

Comment:
Can you add a test that it was Some() before on_init?

Reply:
We already test for this. Immediately above the reinitialization of the patch manager, we verify that next_boot_patch is numbered `1`.

Reply:
True.  I guess I would have tested that *after* deserialization, before init to test that it's actually being serialized.  🤷

Reply:
Ah! Understood. Done.